DOC: Add example to show usage of ``predecessors`` matrix returned from `sparse.csgraph.shortest_path`

czgdp1807 opened the pull request:
<!-- 
Thanks for contributing a pull request! Please ensure that
your PR satisfies the checklist before submitting:
https://scipy.github.io/devdocs/dev/contributor/development_workflow.html#checklist-before-submitting-a-pr

Also, please name and describe your PR as you would write a
commit message:
https://scipy.github.io/devdocs/dev/contributor/development_workflow.html#writing-the-commit-message.
However, please only include an issue number in the description, not the title,
and please ensure that any code names containing underscores are enclosed in backticks.

Depending on your changes, you can skip CI operations and save time and energy: 
https://scipy.github.io/devdocs/dev/contributor/continuous_integration.html#skipping

Note that we are a team of volunteers; we appreciate your
patience during the review process.

Again, thanks for contributing!
-->

#### Reference issue
<!--Example: Closes gh-WXYZ.-->

Closes https://github.com/scipy/scipy/issues/13914

#### What does this implement/fix?
<!--Please explain your changes.-->

In this PR I have added one example to show how to use ``predecessors`` matrix returned from `sparse.csgraph.shortest_path`. This example will help in clearing up how to read the said matrix and use it to re-construct the shortest paths from sources to nodes. https://github.com/scipy/scipy/issues/13914 was opened due to the confusion in the format of ``predecessors`` matrix. The example I have added helps in clearing that up because it uses ``predecessors`` matrix to construct shortest paths (which is what it is meant for).

@bernstei Is this good? 

The documentation of `predecessors` is good enough, the example just adds on it and concretely shows how to read it.

#### Additional information
<!--Any additional information you think is important.-->

czgdp1807 committed:
DOC: Added example to show usage of predecessors matrix for constructing shortest paths

czgdp1807 committed:
MISC: Remove trailing space

czgdp1807 committed:
DOC: Add comments in the loop constructing shortest paths

j-bowhay commented on scipy/sparse/csgraph/_shortest_path.pyx:
```diff
@@ -168,6 +168,60 @@ def shortest_path(csgraph, method='auto',
+    ... [0, 5, 0, 0]
+    ... ]
+    >>> graph = csr_array(graph)
+    >>> print(graph)
```
Let's remove this print, the previous example shows what csr array looks like when printed

j-bowhay commented on scipy/sparse/csgraph/_shortest_path.pyx:
```diff
@@ -168,6 +168,60 @@ def shortest_path(csgraph, method='auto',
+    >>> dist_matrix
+    array([[ 0., 15.,  7., 20.],
+           [ 7.,  8.,  0., 13.]])
+    >>> predecessors
```
Up until this point the example is exactly the same as the old one, we could probably just remove the old one for conciseness?

czgdp1807 replied:
> we could probably just remove the old one for conciseness?

Makes sense. Would prefer removing. The new example is better IMO.

j-bowhay commented on scipy/sparse/csgraph/_shortest_path.pyx:
```diff
@@ -168,6 +168,60 @@ def shortest_path(csgraph, method='auto',
     >>> predecessors
     array([-9999,     0,     0,     1], dtype=int32)
 
+    >>> graph = [
```
It would be great to start the example with a sentence on what I'll show

j-bowhay commented on scipy/sparse/csgraph/_shortest_path.pyx:
```diff
@@ -168,6 +168,60 @@ def shortest_path(csgraph, method='auto',
+    array([[-9999,     2,     0,     1],
+           [    2,     2, -9999,     1]], dtype=int32)
+
+    >>> shortest_paths = {}
```
It would be great to have some text here explaining what this is about to show

j-bowhay commented on scipy/sparse/csgraph/_shortest_path.pyx:
```diff
@@ -168,6 +168,60 @@ def shortest_path(csgraph, method='auto',
+    ...         shortest_paths[(sources[idx], node)] = path
+    ...
+
+    >>> shortest_paths[(0, 3)]
```
Again some commentary here would be great

j-bowhay reviewed (changes requested):
Thanks @czgdp1807, a few comments. In general it would be good to include more commentary in the example

jakevdp reviewed (commented):
The example here would be more useful with further explanation of what the code is intended to do. I think this may be difficult without an example where the graph has more meaning; for instance there's already one example of using the predecessors array in https://docs.scipy.org/doc/scipy/tutorial/csgraph.html#example-word-ladders Maybe linking to that would be sufficient?

czgdp1807 commented:
> Maybe linking to that would be sufficient?

The link you shared is very nice. Linking it is a nice idea and I would do it. However, in my opinion it's one full page to read. The example I added here is very short. With more commentary it will quickly help the users in understanding the semantics of `predecessors`. Let me know if you still want to just have https://docs.scipy.org/doc/scipy/tutorial/csgraph.html#example-word-ladders linked and remove the example I added.

Meanwhile I wait for your response I will work on linking https://docs.scipy.org/doc/scipy/tutorial/csgraph.html#example-word-ladders. Will remove my example once you say so.l

czgdp1807 committed:
DOC: Linked Word Ladders example in shortest_path documentation

czgdp1807 committed:
DOC: Removed redundant example from shortest_path documentation

czgdp1807 committed:
DOC: Add commentary to explain shortest path reconstruction

czgdp1807 committed:
[docs only]
